.github/actions/detect-workflow/main_test.go

```diff
@@ -172,3 +174,67 @@ func Test_action_getRepoRef(t *testing.T) {
 		}
 	})
 }
+
+func Test_newAction(t *testing.T) {
+	type args struct {
+		getenv func(string) string
+		c      *github.OIDCClient
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *action
+		wantErr bool
+	}{
+		{
+			name: "failure with empty string",
+			args: args{
+				getenv: func(k string) string {
+					return ""
+				},
+			},
+			wantErr: true,
+			want:    nil,
+		},
+		{
+			name: "success",
+			args: args{
+				getenv: func(k string) string {
+					// create a temp file with key:value
+					f, err := ioutil.TempFile("", "")
+					if err != nil {
+						t.Fatalf("unexpected error: %v", err)
+					}
+					defer f.Close()
+					if _, err := f.Write([]byte(`{"test":"hoge"}`)); err != nil {
+						t.Fatalf("unexpected error: %v", err)
+					}
+					return f.Name()
+				},
+			},
+			want: &action{
+				getenv: func(k string) string {
+					return "test:hoge"
+				},
+				event: map[string]any{
+					"test": "hoge",
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := newAction(tt.args.getenv, tt.args.c)
+			if (err != nil) != tt.wantErr {
```

Comment:
You can return a specific error value and check for that instead.

Reply:
This would mean refactoring all the errors are wrapping them up into custom errors. There are 3 errors within this small function.

```
func newAction(getenv func(string) string, c *github.OIDCClient) (*action, error) {
	eventPath := getenv("GITHUB_EVENT_PATH")
	if eventPath == "" {
		return nil, errors.New("GITHUB_EVENT_PATH not set")
	}

	payload, err := os.ReadFile(eventPath)
	if err != nil {
		return nil, err
	}

	var event map[string]any
	if err := json.Unmarshal(payload, &event); err != nil {
		return nil, err
	}

	return &action{
		getenv: getenv,
		event:  event,
		client: c,
	}, nil
}
```

This is out of scope for this PR and should be another PR for addressing [errors](https://github.com/polyfloyd/go-errorlint)

Reply:
I see what you mean. I'm ok with addressing it in a later PR, but I thought that since you are improving the tests with this PR that it would be in scope to also modify the function in a way that makes it easier to test. As it is, your test will pass if any error is returned when I think the intention is to check for errors in a specific case. wdut?

Just to elaborate and be more specific about my thoughts, I think that since you are only testing the one case so far, you can address this by modifying the just the first returned error without changing the others. And, since it's not a dynamic error you can return a static error value.

```golang
var errEventPathNotSet = errors.New("GITHUB_EVENT_PATH not set")

func newAction(getenv func(string) string, c *github.OIDCClient) (*action, error) {
    eventPath := getenv("GITHUB_EVENT_PATH")
    if eventPath == "" {
        return nil, errEventPathNotSet
    }
    ...
}
```

and in the test you can set `wantErr` to an error type and check it with `errors.Is`.

```
tests := []struct {
	name    string
	args    args
	want    *action
	wantErr error
}{
    {
        ...
        wantErr: errEventPathNotSet,
     },
     ...
 }

if !errors.Is(err, tt.wantErr) {
    t.Fatalf(...)
}
```

I believe this is generally how we've tried to write tests that check errors throughout the project, using `errors.Is` or `errors.As`. You can see some examples (albiet a bit more complicated) in [oidc.go](https://github.com/slsa-framework/slsa-github-generator/blob/d98cbc192096122d80204d5d602958530fac6162/github/oidc.go#L64-L87) and [oidc_test.go](https://github.com/slsa-framework/slsa-github-generator/blob/d98cbc192096122d80204d5d602958530fac6162/github/oidc_test.go#L62-L64). Another example is [build.go](https://github.com/slsa-framework/slsa-github-generator/blob/d98cbc192096122d80204d5d602958530fac6162/internal/builders/go/pkg/build.go#L27-L33) and [build_test.go](https://github.com/slsa-framework/slsa-github-generator/blob/d98cbc192096122d80204d5d602958530fac6162/internal/builders/go/pkg/build_test.go#L98-L100).

We also have an internal [errors](https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/errors) package to help support this.

I hope that makes my thoughts clearer!